### PR TITLE
Replacing the policy registry with policy "aliases"

### LINKED ISSUE #88: [Bug] Can't use as base policy a policy that is derivative of an existing policy
**Describe the bug**
A derived policy can't be used as the base policy for another algorithm.
Use case:

Derivative algorithm that expects a specific kind of policy.

**Code example**
```python
from typing import List, Tuple, Type, Union, Callable, Optional, Dict, Any
import numpy as np
import torch as th
import torch.nn.functional as F
from stable_baselines3.common import logger
from stable_baselines3.common.off_policy_algorithm import OffPolicyAlgorithm
from stable_baselines3.common.type_aliases import GymEnv, MaybeCallback
from stable_baselines3.common.utils import get_linear_fn
from stable_baselines3.dqn import DQN
from stable_baselines3.common.policies import register_policy
from stable_baselines3.dqn.policies import DQNPolicy

class SpecialDQNPolicy(DQNPolicy):
    def __init__(*vargs, **kwargs):
        super().__init__(*vargs, **kwargs)

SpecialMlpPolicy = SpecialDQNPolicy
register_policy("SpecialMlpPolicy", SpecialMlpPolicy)


class SpecialDQN(DQN):
    def __init__(self, policy: Union[str, Type[SpecialDQNPolicy]],
                 env: Union[GymEnv, str],
                 learning_rate: Union[float, Callable] = 1e-4,
                 buffer_size: int = 1000000,
                 learning_starts: int = 50000,
                 batch_size: Optional[int] = 32,
                 tau: float = 1.0,
                 gamma: float = 0.99,
                 train_freq: int = 4,
                 gradient_steps: int = 1,
                 n_episodes_rollout: int = -1,
                 optimize_memory_usage: bool = False,
                 target_update_interval: int = 10000,
                 exploration_fraction: float = 0.1,
                 exploration_initial_eps: float = 1.0,
                 exploration_final_eps: float = 0.05,
                 max_grad_norm: float = 10,
                 tensorboard_log: Optional[str] = None,
                 create_eval_env: bool = False,
                 policy_kwargs: Optional[Dict[str, Any]] = None,
                 verbose: int = 0,
                 seed: Optional[int] = None,
                 device: Union[th.device, str] = 'auto',
                 _init_setup_model: bool = True):
        super(DQN, self).__init__(policy, env, SpecialDQNPolicy, learning_rate,
                                  buffer_size, learning_starts, batch_size,
                                  tau, gamma, train_freq, gradient_steps,
                                  n_episodes_rollout, action_noise=None,  # No action noise
                                  policy_kwargs=policy_kwargs,
                                  tensorboard_log=tensorboard_log,
                                  verbose=verbose, device=device,
                                  create_eval_env=create_eval_env,
                                  seed=seed, sde_support=False,
                                  optimize_memory_usage=optimize_memory_usage)
        self.exploration_initial_eps = exploration_initial_eps
        self.exploration_final_eps = exploration_final_eps
        self.exploration_fraction = exploration_fraction
        self.target_update_interval = target_update_interval
        self.max_grad_norm = max_grad_norm
        # "epsilon" for the epsilon-greedy exploration
        self.exploration_rate = 0.0
        # Linear schedule will be defined in `_setup_model()`
        self.exploration_schedule = None
        self.q_net, self.q_net_target = None, None
        if _init_setup_model:
            self._setup_model()


```
Error
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 35, in __init__
  File "/Users/steliostymvios/Projects/stable-baselines3/stable_baselines3/common/off_policy_algorithm.py", line 100, in __init__
    seed=seed, use_sde=use_sde, sde_sample_freq=sde_sample_freq)
  File "/Users/steliostymvios/Projects/stable-baselines3/stable_baselines3/common/base_class.py", line 70, in __init__
    self.policy_class = get_policy_from_name(policy_base, policy)
  File "/Users/steliostymvios/Projects/stable-baselines3/stable_baselines3/common/policies.py", line 653, in get_policy_from_name
    print(_policy_registry[base_policy_type])
KeyError: <class '__main__.SpecialDQNPolicy'>
```

The registry isn't filled correctly
```
>>> pprint(_policy_registry)
{<class 'stable_baselines3.common.policies.ActorCriticPolicy'>: {'CnnPolicy': <class 'stable_baselines3.common.policies.ActorCriticCnnPolicy'>,
                                                                 'MlpPolicy': <class 'stable_baselines3.common.policies.ActorCriticPolicy'>},
 <class 'stable_baselines3.sac.policies.SACPolicy'>: {'CnnPolicy': <class 'stable_baselines3.sac.policies.CnnPolicy'>,
                                                      'MlpPolicy': <class 'stable_baselines3.sac.policies.SACPolicy'>},
 <class 'stable_baselines3.td3.policies.TD3Policy'>: {'CnnPolicy': <class 'stable_baselines3.td3.policies.CnnPolicy'>,
                                                      'MlpPolicy': <class 'stable_baselines3.td3.policies.TD3Policy'>},
 <class 'stable_baselines3.dqn.policies.DQNPolicy'>: {'CnnPolicy': <class 'stable_baselines3.dqn.policies.CnnPolicy'>,
                                                      'MlpPolicy': <class 'stable_baselines3.dqn.policies.DQNPolicy'>,
                                                      'SpecialCnnPolicy': <class '__main__.SpecialCnnPolicy'>,
                                                      'SpecialMlpPolicy': <class '__main__.SpecialDQNPolicy'>}}
```

### LINKED ISSUE #841: [Question] Refactoring/removing the "policy registry"
### Question

The policy registry is, in my opinion, confusing to understand the codebase and doesn't solve complex problems. I suggest each algorithm simply declare its own aliases in-class instead (I will attach a PR to this)

### Additional context

Related issue: #88 

### Checklist

- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**)
- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)

### PR DESCRIPTION
**Pull request by Gregwar**
## Description

Here is a proposal of simplification of the "policy registry" system. It basically removes the policy registry and `register_policy` in favour of a `policy_aliases` dict that is defined by algorithms.

**This PR is not complete and should not be merged without extra discussions. It also needs to be reflected in SB3 contrib.**

## Motivation and Context

- I think this registry system is more complex and confusing than it should (see issue #88 for instance)
- At a first glance at the code, you need to dive into this to understand where policies are built, while I think this proposal makes it clearer
- Some `policies.py` files could be arguably removed (e.g: a2c's, in favour of directly specifying the names in `a2c.py`)
- Needs also to be reflected in contribs to keep compatibility

- [X] See #841 - I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)
- I didn't personally raised an issue, but #88 is an example of such confusion

closes #88
closes #841

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:

- [ ] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [X] I have updated the changelog accordingly (**required**).
- [X] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [X] I have reformatted the code using `make format` (**required**)
- [X] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [X] I have ensured `make pytest` and `make type` both pass. (**required**)
- [ ] I have checked that the documentation builds using `make doc` (**required**)

### CODE REVIEW

**Comment by Gregwar** [comment]
Changing import order create a circular import error, not sure why (and whether this should be muted/ignored or fixed)

**Comment by araffin** [comment]
> Changing import order create a circular import error, not sure why (and whether this should be muted/ignored or fixed)

checks are fine on my side, make sure to use latest version of isort and black (and to use `make commit-checks`).
Careful with some mutable as default...

File: stable_baselines3/common/base_class.py [hunk 1]
```diff
@@ -325,6 +326,23 @@ def _excluded_save_params(self) -> List[str]:
...
+        and they receive respective policies that work for them.
+
+        :return:
+            A policy class (type)
+        """
+
+        if policy_name in self.policy_aliases:
+            return self.policy_aliases[policy_name]
+        else:
+            raise ValueError("Policy %s unknown" % policy_name)
```
Comment (araffin) [inline stable_baselines3/common/base_class.py:344]:
```suggestion
            raise ValueError(f"Policy {policy_name} unknown")
```

Please use f-strings

File: stable_baselines3/common/off_policy_algorithm.py [hunk 2]
```diff
@@ -96,6 +95,7 @@ def __init__(
         support_multi_env: bool = False,
         create_eval_env: bool = False,
         monitor_wrapper: bool = True,
+        policy_aliases: Dict[str, Type[BasePolicy]] = {},
```
Comment (araffin) [inline stable_baselines3/common/off_policy_algorithm.py:98]: mutable as default is evil in python ;)
https://github.com/satwikkansal/wtfpython#-beware-of-default-mutable-arguments

**Comment by araffin** [comment]
> create a circular import error

ah, you mean for the doc. Let me check that.

**Comment by Gregwar** [comment]
> @Gregwar should be fixed now ;)

Where did you fixed that ?

**Comment by araffin** [comment]
> Where did you fixed that ?

Was fixed in https://github.com/DLR-RM/stable-baselines3/pull/842/commits/bae5e2103b3f6bc8a27975ad1522b17a0e16cb31


btw, I don't know how you did that, but you reintroduced the deleted `remove_time_limit_termination` parameter (you probably started working on this PR from an old version of master)

**Comment by Gregwar** [comment]
I am sorry something wrong happened indeed, I am fixing this and I will review the files changed by the PR to check that only intended changes are covered here

File: stable_baselines3/a2c/a2c.py [hunk 3]
```diff
@@ -114,6 +114,12 @@ def __init__(
         if _init_setup_model:
             self._setup_model()
 
+    policy_aliases: Dict[str, Type[BasePolicy]] = {
+        "MlpPolicy": ActorCriticCnnPolicy,
```
Comment (araffin) [inline stable_baselines3/a2c/a2c.py:118]: typo, should not be Cnn, test will fail ;)

**Comment by Gregwar** [comment]
I can open a dual PR on the contrib repository if you want

**Comment by araffin** [comment]
> I can open a dual PR on the contrib repository if you want

I was about to ask, please do ;)

File: stable_baselines3/a2c/a2c.py [hunk 4]
```diff
@@ -114,6 +114,12 @@ def __init__(
         if _init_setup_model:
             self._setup_model()
 
+    policy_aliases: Dict[str, Type[BasePolicy]] = {
```
Comment (araffin) [inline stable_baselines3/a2c/a2c.py:117]: is there a reason for defining it after the `_init_` and not before?

Comment (Gregwar) [reply stable_baselines3/a2c/a2c.py:117]: None, I hesitated
Would defining it before `__init__` imply defining it before the class's docstring ?

Comment (araffin) [reply stable_baselines3/a2c/a2c.py:117]: I would actually define it right after the class docstring (if possible)

Comment (Gregwar) [reply stable_baselines3/a2c/a2c.py]: Done

**Comment by araffin** [comment]
So, only the changelog is missing?

**Comment by Gregwar** [comment]
I added a line in the changelog

**Review by araffin — APPROVED** [review]
LGTM, thanks =)

**Pull request merged**